test/pxd_fastpath_test.cc

```diff
@@ -0,0 +1,927 @@
+		oh.unique = 0;
+		oh.error = PXD_REMOVE;
+		oh.len = sizeof(oh) + sizeof(remove);
+
+		remove.dev_id = dev_id;
+		remove.force = false; //// cannot force
+
+		iov[0].iov_base = &oh;
+		iov[0].iov_len = sizeof(oh);
+		iov[1].iov_base = &remove;
+		iov[1].iov_len = sizeof(remove);
+
+		ssize_t write_bytes = writev(ctl_fd, iov, 2);
+		if (write_bytes > 0) {
+			fprintf(stderr, "device removal success\n");
+			ASSERT_EQ(write_bytes, oh.len);
+			break;
+		}
+
+		ASSERT_EQ(EBUSY, errno);
+		fprintf(stderr, "device busy.. will retry after sleep\n");
+		iter++;
+		sleep(1);
+	}
+
+	fprintf(stderr, "%s: device %ld removed after %d secs\n", __func__, dev_id, iter);
+	fprintf(stderr, "prepping to stop background cleaner\n");
+	killed = true;
+	sleep(1);
+	cleaner.join();
+	killed = false;
+}
+
+int PxdFastpathTest::wait_msg(int timeout)
+{
+    struct pollfd fds = {};
+    int ret;
+
+    fds.fd = ctl_fd;
+    fds.events = POLLIN;
+    ret = poll(&fds, 1, timeout * 1000);
+    if (ret > 0)
+		return 0;
+	if (ret == 0)
+		return -ETIMEDOUT;
+
+	// should never arise?!
+	ret = -errno;
+	EXPECT_GE(ret, 0);
+	return ret;
+}
+
+void PxdFastpathTest::read_block(fuse_in_header *in, pxd_rdwr_in *rd)
+{
+    auto buf = aligned_buffer_fastpath(rd->size);
+    init_pattern_fastpath(buf.get(), rd->size);
+
+    fuse_out_header oh;
+    struct iovec wr_iov[3];
+    int ret;
+
+    oh.unique = in->unique;
+    oh.error = 0;
+    oh.len = sizeof(oh) + rd->size;
+
+    wr_iov[0].iov_base = &oh;
+    wr_iov[0].iov_len = sizeof(oh);
+    wr_iov[1].iov_base = buf.get();
+    wr_iov[1].iov_len = rd->size;
+
+    // Send a read response to kernel
+    ret = writev(ctl_fd, wr_iov, 2);
+    fprintf(stderr, "%s: sent read response to kernel\n", __func__);
+    ASSERT_EQ(ret, oh.len);
+}
+
+void PxdFastpathTest::validate_device_properties(const std::string &device_name,
+                                                 uint64_t expected_discard_granularity,
+                                                 uint64_t expected_max_discard_bytes)
+{
+    // Check if device exists
+    ASSERT_TRUE(access(device_name.c_str(), F_OK) == 0)
+        << "Device " << device_name << " does not exist";
+
+    // Read sysfs attributes to validate device properties
+    std::string dev_name = device_name.substr(device_name.find_last_of('/') + 1);
+    std::string sysfs_base = "/sys/block/pxd!" + dev_name;
+
+    // Check queue directory exists
+    std::string queue_dir = sysfs_base + "/queue";
+    ASSERT_TRUE(access(queue_dir.c_str(), F_OK) == 0)
+        << "Queue directory " << queue_dir << " does not exist";
+
+    std::cout << "Device properties validated for: " << device_name << std::endl;
+}
+
+void PxdFastpathTest::validate_fastpath_active(const std::string &device_name)
+{
+    // Extract device minor number from device name
+    std::string dev_name = device_name.substr(device_name.find_last_of('/') + 1);
+    std::string minor_str = dev_name.substr(3); // Remove "pxd" prefix
```

Comment:
minor number is returned when device is attached (along with io path status (fastpath vs nativepath)) use it instead.